sdk/tests/deqp/resources/deqp-draw.js

```diff
@@ -0,0 +1,253 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+
+var deqpDraw = (function() {
+'use strict';
+
+/**
+ * Description of a vertex array binding
+ * @param {WebGLRenderingContext.GLEnum} type GL Type of data
+ * @param {string|number} location Binding location
+ * @param {number} components Number of components per vertex
+ * @param {number} elements Number of elements in the array
+ * @param {Array.<number>} data Source data
+ */
+var VertexArrayBinding = function(type, location, components, elements, data) {
+    this.type = type;
+    this.location = location;
+    this.components = components;
+    this.elements = elements;
+    this.data = data;
+};
+
+/**
+ * Creates vertex buffer, binds it and draws elements
+ * @param {WebGLRenderingContext} gl WebGL context
+ * @param {number} program ID, vertexProgramID
+ * @param {Array.<number>} vertexArrays
+ * @param {PrimitiveList} primitives to draw
+ * @param {function()} callback
+ */
+var draw = function(gl, program, vertexArrays, primitives, callback) {
+    /** TODO: finish implementation */
+    /** @type {Array.<WebGLBuffer>} */ var objects = [];
+
+    for (var i = 0; i < vertexArrays.length; i++) {
+        /** @type {WebGLBuffer} */ var buffer = vertexBuffer(gl, vertexArrays[i]);
+        objects.push(buffer);
+    }
+
+    if (primitives.indices) {
+        /** @type {WebGLBuffer} */ var elemBuffer = indexBuffer(gl, primitives);
+        gl.bindBuffer(gl.ELEMENT_ARRAY_BUFFER, elemBuffer);
+
+        if (callback)
+            callback.beforeDrawCall();
+
+        drawIndexed(gl, primitives, 0);
+
+        if (callback)
+            callback.afterDrawCall();
+
+        gl.bindBuffer(gl.ELEMENT_ARRAY_BUFFER, null);
+    } else {
+    /** TODO: implement */
+    }
+
+  assertMsgOptions(gl.getError() === gl.NO_ERROR, 'drawArrays', false, true);
+    for (var i = 0; i < vertexArrays.length; i++) {
+        gl.disableVertexAttribArray(vertexArrays[i].location);
+    }
+  gl.bindBuffer(gl.ARRAY_BUFFER, null);
+};
+
+/**
+ * Creates vertex buffer, binds it and draws elements
+ * @param {WebGLRenderingContext} gl WebGL context
+ * @param {PrimitiveList} primitives Primitives to draw
+ * @param {number} offset
+ */
+var drawIndexed = function(gl, primitives, offset) {
+/** @type {WebGLRenderingContext.GLEnum} */ var mode = getPrimitiveGLType(gl, primitives.type);
+    /** TODO: C++ implementation supports different index types, we use only int16.
+        Could it cause any issues?
+
+        deUint32 indexGLType = getIndexGLType(primitives.indexType);
+    */
+
+    gl.drawElements(mode, primitives.indices.length, gl.UNSIGNED_SHORT, offset);
+};
+
+
+/**
+ * Enums for primitive types
+ * @enum
+ */
+var primitiveType = {
+    TRIANGLES: 0,
+    TRIANGLE_STRIP: 1,
+    TRIANGLE_FAN: 2,
+
+    LINES: 3,
+    LINE_STRIP: 4,
+    LINE_LOOP: 5,
+
+    POINTS: 6,
+
+    PATCHES: 7
```

Comment:
The PATCHES primitive type doesn't exist in OpenGL ES 2.0 or WebGL. Should be removed so the code below throws an exception if somebody tries to use it.